From 56c0f28f43e38d29a3fc8513dba959174d91dcef Mon Sep 17 00:00:00 2001
From: Huynh Thanh Hung <hung.huynh.wz@renesas.com>
Date: Thu, 6 Jun 2019 10:16:49 +0700
Subject: [PATCH] Fix division by zero in radial gradiants with NEON

The NEON implementation uses rsqrt and thus can not be taken on 0, so
replace the minimum with something close to zero instead of zero.

Task-number: QTBUG-59961 - detected from upstream.

Reviewed-by: Erik Verbruggen <erik.verbruggen@qt.io>
Signed-off-by: Huynh Thanh Hung <hung.huynh.wz@renesas.com>
---
 src/gui/painting/qdrawhelper_p.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/gui/painting/qdrawhelper_p.h b/src/gui/painting/qdrawhelper_p.h
index a6d1286..5c1c3a9 100644
--- a/src/gui/painting/qdrawhelper_p.h
+++ b/src/gui/painting/qdrawhelper_p.h
@@ -527,7 +527,12 @@ public:
         const typename Simd::Float32x4 v_r0 = Simd::v_dup(data->gradient.radial.focal.radius);
         const typename Simd::Float32x4 v_dr = Simd::v_dup(op->radial.dr);
 
+#if defined(__ARM_NEON__)
+        // NEON doesn't have SIMD sqrt, but uses rsqrt instead that can't be taken of 0.
+        const typename Simd::Float32x4 v_min = Simd::v_dup(std::numeric_limits<float>::epsilon());
+#else
         const typename Simd::Float32x4 v_min = Simd::v_dup(0.0f);
+#endif
         const typename Simd::Float32x4 v_max = Simd::v_dup(float(GRADIENT_STOPTABLE_SIZE-1));
         const typename Simd::Float32x4 v_half = Simd::v_dup(0.5f);
 
-- 
2.7.4

